File: LeetCode/88-MergeSortedArray/GO/main.go
package main

import "fmt"

func main() {
	nums1 := []int{1, 2, 3, 0, 0, 0}
	m := 3
	nums2 := []int{2, 5, 6}
	n := 3
	merge(nums1, m, nums2, n)
	fmt.Println(nums1)
}

func merge(nums1 []int, m int, nums2 []int, n int) {
	last := m + n - 1
	m -= 1
	n -= 1

	for m >= 0 && n >= 0 {
		if nums1[m] > nums2[n] {
			nums1[last] = nums1[m]
			m -= 1
		} else {
			nums1[last] = nums2[n]
			n -= 1
		}
		last -= 1
	}
	for n >= 0 {
		nums1[last] = nums2[n]
		last -= 1
		n -= 1
	}
}

File: LeetCode/88-MergeSortedArray/Rust/main.rs
struct Solution {}
impl Solution {
    pub fn merge(nums1: &mut Vec<i32>, m: i32, nums2: &mut Vec<i32>, n: i32) {
        let (mut m, mut n) = (m as usize, n as usize);
        let mut last = m + n - 1;
        while m > 0 && n > 0 {
            if nums1[last] > nums2[m - 1] {
                nums1[last] = nums1[m - 1];
                m -= 1;
            } else {
                nums1[last] = nums2[n - 1];
                n -= 1;
            }
            last -= 1
        }
        while n > 0 {
            nums1[last] = nums2[n - 1];
            n -= 1;
            last -= 1;
        }
    }
}

fn main() {
    let mut nums1 = vec![1, 2, 3, 0, 0, 0];
    let m = 3;
    let mut nums2 = vec![2, 5, 6];
    let n = 3;
    Solution::merge(&mut nums1, m, &mut nums2, n);
    println!("{:?}", nums1);
}
